ckanext/canada/view.py

```diff
@@ -813,125 +803,79 @@ def _org_key(org):
-    try:
-        function = get_action(logic_function)
-    except KeyError:
-        log.info('Can\'t find logic function: %s', logic_function)
-        return _finish_bad_request(
-            _('Action name not known: %s') % logic_function)
-
-    context = {'model': model, 'session': model.Session, 'user': g.user,
-                'api_version': ver, 'auth_user_obj': g.userobj}
-    model.Session()._context = context
-
-    return_dict = {'help': h.url_for('api.action',
-                                        logic_function='help_show',
-                                        ver=ver,
-                                        name=logic_function,
-                                        qualified=True)}
-    try:
-        side_effect_free = getattr(function, 'side_effect_free', False)
-        request_data = _get_request_data(try_url_params=
-                                                side_effect_free)
-    except ValueError as inst:
-        log.info('Bad Action API request data: %s', inst)
-        return _finish_bad_request(
-            _('JSON Error: %s') % inst)
-    if not isinstance(request_data, dict):
-        # this occurs if request_data is blank
-        log.info('Bad Action API request data - not dict: %r',
-                    request_data)
-        return _finish_bad_request(
-            _('Bad request data: %s') %
-            'Request data JSON decoded to %r but '
-            'it needs to be a dictionary.' % request_data)
-    # if callback is specified we do not want to send that to the search
-    if 'callback' in request_data:
-        del request_data['callback']
-        g.user = None
-        g.userobj = None
-        context['user'] = None
-        context['auth_user_obj'] = None
+    u'''Main endpoint for the action API (v3)
+
+    Creates a dict with the incoming request data and calls the appropiate
+    logic function. Returns a JSON response with the following keys:
+
+        * ``help``: A URL to the docstring for the specified action
+        * ``success``: A boolean indicating if the request was successful or
+                an exception was raised
+        * ``result``: The output of the action, generally an Object or an Array
+
+    Canada Fork:
+        We keep version 1 and 2 endpoints just incase any systems are still using that.
+        We also have -1 to version to return the context and request_data for extra logging.
+        And if the request is a POST request, we want to not authorize any PD type.
+    '''
+    context, request_data, return_dict = api_view_action(logic_function, ver=-1)
+
+    # extra logging here
+    id = request_data.get('id', request_data.get('package_id', request_data.get('resource_id')))
+    pkg_dict = _get_package_from_api_request(logic_function, id, context)
+    if pkg_dict:
+        _log_api_access(request_data, pkg_dict)
+
+    # prevent PD types from being POSTed to via the API
+    if request.method == 'POST':
```

Comment:
We want users to be able to upsert/delete PD records with the API, won't this change prevent that?

Reply:
From testing it, it seems that you are able to still use all of the `DataStore` endpoints. Which is what the users will use to manage the PD data. I am not sure why this is even possible, because yeah, from the look of this new code it SHOULD prevent that

Reply:
it must not be able to find the package or something. Will be debugging and seeing. I can just make sure that the actions are not a `datastore` one too

Reply:
@wardi yeah when getting the pkg_dict, I had `if logic_function.startswith('resource')` and then did `resource_show` in that block. But of course the logic actions start with `datastore`. So conditions have been updated.

resource and package action endpoints will `403` for PD types. But datastore action endpoints are allowed.